File: arrays_and_strings/reverse_string.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <vector>

//! @brief First attempt solution
//! @param[in,out] s Vector of chars to reverse
static void reverseStringFA(std::vector<char>& s)
{
    if (s.empty())
    {
        return;
    }

    int  left {};
    auto right = static_cast<int>(s.size()) - 1;

    while (left < right)
    {
        std::swap(s[left], s[right]);
        ++left;
        --right;
    }

} // static void reverseStringFA( ...

//! @brief First discussion solution
//! @param[in,out] s Vector of chars to reverse
static void reverseStringDS1(std::vector<char>& s)
{
    std::reverse(s.begin(), s.end());
}

TEST(ReverseStringTest, SampleTest)
{
    const std::vector<char> input {'h', 'e', 'l', 'l', 'o'};
    const std::vector<char> expected_output {'o', 'l', 'l', 'e', 'h'};

    auto s_fa = input;
    reverseStringFA(s_fa);
    EXPECT_TRUE(std::equal(expected_output.cbegin(),
                           expected_output.cend(),
                           s_fa.cbegin()));
    
    auto s_ds1 = input;
    reverseStringDS1(s_ds1);
    EXPECT_TRUE(std::equal(expected_output.cbegin(),
                           expected_output.cend(),
                           s_ds1.cbegin()));
}

File: arrays_and_strings/reverse_string.py
def reverseStringShort(s: List[str]) -> None:
    s.reverse()


def reverseStringRecursive(s: List[str]) -> None:
    def helper(left: int, right: int) -> None:
        if left < right:
            s[left], s[right] = s[right], s[left]
            helper(left + 1, right - 1)

    helper(0, len(s) - 1)


def reverseStringTwoPointers(s: List[str]) -> None:
    left, right = 0, len(s) - 1
    while left < right:
        s[left], s[right] = s[right], s[left]
        left, right = left + 1, right - 1

